tortoise/verifying_tortoise.go

```diff
@@ -291,243 +371,233 @@ func RecoverVerifyingTortoise(mdb retriever) (interface{}, error) {
 
-		// Count good blocks votes..
-		// Declare the vote vector “verified” up to position k if the total weight exceeds the confidence threshold in all positions up to k .
-		for blk, vote := range input {
+		// Count the votes of good blocks. vote is our opinion on this block.
+		// Declare the vote vector “verified” up to position k if the total weight exceeds the confidence threshold in
+		// all positions up to k
+		for blk, vote := range layerVotes {
 			// Count the votes for the input vote vector by summing the weight of the good blocks
 			sum := abstain
-			//t.logger.Info("counting votes for block %v", blk)
-			for j := t.Last; j > i && j-i < t.Hdist; j-- {
+
+			// Step backwards from the latest received layer to i+1, but only count votes for the interval [i, i+Hdist]
+			var startLayer types.LayerID
+			if t.Last < i+t.Hdist {
+				startLayer = t.Last
+			} else {
+				startLayer = i + t.Hdist - 1
+			}
+			for j := startLayer; j > i; j-- {
 				// check if the block is good
-				for bid, op := range t.BlocksToBlocks[j] {
-					_, isgood := t.GoodBlocksIndex[bid]
-					if !isgood {
-						t.logger.With().Debug("block not good hence not counting", log.FieldNamed("voting_block", bid), log.FieldNamed("voted_block", blk))
+				for bid, op := range t.BlockOpinionsByLayer[j] {
+					if _, isgood := t.GoodBlocksIndex[bid]; !isgood {
+						t.logger.With().Debug("not counting vote of block not marked good",
+							log.FieldNamed("voting_block", bid),
+							log.FieldNamed("voted_block", blk))
 						continue
 					}
 
-					// check if this block has Opinion on this block. (TODO: maybe doesn't have Opinion means AGAINST?)
+					// check if this block has an opinion on this block (TODO: maybe no opinion means AGAINST?)
 					opinionVote, ok := op.BlocksOpinion[blk]
 					if !ok {
-						t.logger.With().Debug("no opinion on block", log.FieldNamed("voting_block", bid), log.FieldNamed("voted_block", blk))
+						t.logger.With().Debug("no opinion on block",
+							log.FieldNamed("voting_block", bid),
+							log.FieldNamed("voted_block", blk))
 						continue
 					}
 
-					t.logger.With().Debug("adding block opinion to vote sum", log.FieldNamed("voting_block", bid), log.FieldNamed("voted_block", blk),
-						log.String("vote", opinionVote.String()), log.String("sum", fmt.Sprintf("[%v, %v]", sum[0], sum[1])))
+					t.logger.With().Debug("adding block opinion to vote sum",
+						log.FieldNamed("voting_block", bid),
+						log.FieldNamed("voted_block", blk),
+						log.String("vote", opinionVote.String()),
+						log.String("sum", fmt.Sprintf("[%v, %v]", sum[0], sum[1])))
 					sum = sum.Add(opinionVote.Multiply(t.BlockWeight(bid, blk)))
 				}
 			}
 
 			// check that the total weight exceeds the confidence threshold in all positions up
-			threshold := float64(globalThreshold*float64(i-wasVerified)) * float64(t.AvgLayerSize)
+			threshold := globalThreshold * float64(i-pbaseOld) * float64(t.AvgLayerSize)
```

Comment:
basically this line is here only for the log. we can't log it in `calculateGlobalOpinion` since it is not a method of turtle and has no access to `logger`..

Reply:
I'll pass the logger in to `calculateGlobalOpinion` so as to avoid the duplication of logic here - good catch